docs/fundamentals/flow/client.md

```diff
@@ -0,0 +1,332 @@
+    ...  # process docs here
+```
+
+````
+
+In the example below, our Flow passes the message then prints the result when successful.
+If something goes wrong, it beeps. Finally, the result is written to output.txt.
+
+```python
+from jina import Flow, Client
+from docarray import Document
+
+
+def beep(*args):
+    # make a beep sound
+    import sys
+
+    sys.stdout.write('\a')
+
+
+with Flow().add() as f, open('output.txt', 'w') as fp:
+    client = Client(port=f.port_expose)
+    client.post(
+        '/',
+        Document(),
+        on_done=print,
+        on_error=beep,
+        on_always=lambda x: x.docs.save(fp),
+    )
+```
+
+## Returning results from .post()
+
+If no callback is provided, `client.post()` returns a flattened `DocumentArray` containing all Documents of all Requests.
+
+By setting `return_responses=True` when creating a Client, this behavior can be modified to return a list of Responses
+(`DataRequest`s) instead.
+
+If a callback is provided, no results will be returned.
+
+```{admonition} Danger
+:class: danger
+Not using a callback function and instead returning results can come with a **serious performance penalty**.
+
+Callbacks operate on each individual Request, which represents a batch of the data.
+In contrast, returning results requires the accumulation of all results of all Requests.
+This means that you will not receive results until all Requests have been processed.
+This may not only be slower, but also require more memory.
+```
+
+````{tab} Returning DocumentArray
+
+```python
+from jina import Flow, Client
+from docarray import Document
+
+with Flow() as f:
+    client = Client(port=f.port_expose)
+    docs = client.post(on='', inputs=Document(text='Hi there!'))
+    print(docs)
+    print(docs.texts)
+```
+```  
+>>> <DocumentArray (length=1) at 140619524357664>
+>>> ['Hi there!']
+```
+
+````
+````{tab} Returning Responses
+
+```python
+from jina import Flow, Client
+from docarray import Document
+
+with Flow() as f:
+    client = Client(port=f.port_expose, return_responses=True)
+    resp = client.post(on='', inputs=Document(text='Hi there!'))
+    print(resp)
+    print(resp[0].docs.texts)
+```
+``` 
+>>> [<jina.types.request.data.DataRequest ('header', 'parameters', 'routes', 'data') at 140619524354592>]
+>>> ['Hi there!']
+```
+
+````
+````{tab} Using callback function
+
+```python
+from jina import Flow, Client
+from docarray import Document
+
+with Flow() as f:
+    client = Client(port=f.port_expose)
+    resp = client.post(
+        on='',
+        inputs=Document(text='Hi there!'),
+        on_done=lambda resp: print(resp.docs.texts),
+    )
+    print(resp)
+```
+```
+>>> ['Hi there!']
+>>> None
+```
+
+````
+
+## Async Python Client
+
+There also exists an async version of the Python Client so that it can easily be used from an `asyncio` context.
+
+While the standard `Client` is also asynchronous under the hood, its async version exposes this fact to the outside world,
```

Comment:
what is meant by this?

Reply:
I would mention, it allows to `asyncrhonously iterate over responses as they come`

Reply:
> what is meant by this?

That the async version doesn't hide its async-ness, unlike the normal client. To be honest, I don't know all the details about this async client myself, i just kept this more or less unchanged from Tobi's version. Changed it slightly, let me know if it makes more sense to you.

Reply:
Otherwise feel free to suggest and apply changes yourself of course

Reply:
@jacobowitz what is meant with client also being async?